sqlx-core/src/postgres/listen.rs

```diff
@@ -0,0 +1,302 @@
+        loop {
+            // Ensure we have an active connection to work with.
+            let mut conn = match &mut self.connection {
+                Some(conn) => conn,
+                None => return None, // This will never practically be hit, but let's make Rust happy.
+            };
+            // Ensure the current connection has properly registered all listener channels.
+            if self.needs_to_send_listen_cmd {
+                if let Err(err) = send_listen_query(&mut conn, &self.channels).await {
+                    // If we've encountered an error here, test the connection. If the connection
+                    // is good, we return the error. Else, we return `None` as the connection is dead.
+                    if let Err(_) = conn.ping().await {
+                        return None;
+                    }
+                    return Some(Err(err));
+                }
+                self.needs_to_send_listen_cmd = false;
+            }
+            // Await a notification from the DB.
+            match conn.receive().await {
+                // We've received an async notification, return it.
+                Ok(Some(Message::NotificationResponse(notification))) => {
+                    return Some(Ok(notification.into()))
+                }
+                // Protocol error, return the error.
+                Ok(Some(msg)) => {
+                    return Some(Err(protocol_err!(
+                        "unexpected message received from database {:?}",
+                        msg
+                    )
+                    .into()))
+                }
+                // The connection is dead, return None.
+                Ok(None) => return None,
+                // An error has come up, return it.
+                Err(err) => return Some(Err(err)),
+            }
+        }
+    }
+
+    /// Consume this listener, returning a `Stream` of notifications.
+    pub fn stream(mut self) -> impl Stream<Item = Result<PgNotification>> {
+        use async_stream::stream;
+        stream! {
+            loop {
+                match self.recv().await {
+                    Some(res) => yield res,
+                    None => break,
+                }
+            }
+        }
+    }
+}
+
+impl<C> PgListener<C>
+where
+    C: Connection,
+{
+    /// Close this listener stream and its underlying connection.
+    pub async fn close(self) -> BoxFuture<'static, Result<()>> {
+        match self.connection {
+            Some(conn) => conn.close(),
+            None => Box::pin(futures_util::future::ok(())),
+        }
+    }
+}
+
+/// An asynchronous message sent from the database.
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct PgNotification {
+    /// The PID of the database process which sent this notification.
+    pub pid: u32,
+    /// The channel of the notification, which can be thought of as a topic.
+    pub channel: String,
+    /// The payload of the notification.
+    pub payload: String,
+}
```

Comment:
@mehcode you mentioned some stuff about the structure of this guy in an earlier review. Not sure if this is what you wanted to see. If not, we could just wrap the `Box<NotificationResponse>` and provide accessors.